internal/impl/opensearch/output.go

```diff
@@ -0,0 +1,454 @@
+}
+
+// Close shuts down the OpenSearch writer and stops processing messages.
+func (e *OpenSearch) Close(context.Context) error {
+	return nil
+}
+
+// Build a bulkable request for a given pending bulk index item.
+func (e *OpenSearch) buildBulkableRequest(p *pendingBulkIndex) (*opensearchutil.BulkIndexerItem, error) {
+	switch p.Action {
+	case "update":
+		// opensearch util needs the body root element should be "doc"
+		doc, _ := sjson.SetOptions("", "doc", p.Doc, &sjson.Options{ReplaceInPlace: true})
+		r := &opensearchutil.BulkIndexerItem{
+			Index:  p.Index,
+			Action: "update",
+			Body:   strings.NewReader(doc),
+		}
+		if p.ID != "" {
+			r.DocumentID = p.ID
+		}
+		if p.Routing != "" {
+			r.Routing = &p.Routing
+		}
+
+		return r, nil
+	case "delete":
+		r := &opensearchutil.BulkIndexerItem{
+			Index:      p.Index,
+			DocumentID: p.ID,
+			Action:     "delete",
+		}
+		if p.Routing != "" {
+			r.Routing = &p.Routing
+		}
+		return r, nil
+	case "index":
+		jsonData, _ := json.Marshal(p.Doc)
+		r := &opensearchutil.BulkIndexerItem{
+			Index:  p.Index,
+			Action: "index",
+			Body:   bytes.NewReader(jsonData),
+		}
+		if p.ID != "" {
+			r.DocumentID = p.ID
+		}
+		if p.Routing != "" {
+			r.Routing = &p.Routing
+		}
+		return r, nil
+	case "create":
+		jsonData, _ := json.Marshal(p.Doc)
+		r := &opensearchutil.BulkIndexerItem{
+			Index:  p.Index,
+			Action: "create",
+			Body:   bytes.NewReader(jsonData),
+		}
+		if p.ID != "" {
+			r.DocumentID = p.ID
+		}
+		if p.Routing != "" {
+			r.Routing = &p.Routing
+		}
+		return r, nil
+	default:
+		return nil, fmt.Errorf("opensearch action '%s' is not allowed", p.Action)
+	}
+}
+
+// RoundTrip uses the underlying RoundTripper transport, but signs request first with AWS V4 Signing
+func (st Transport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if h, ok := req.Header["Authorization"]; ok && len(h) > 0 && strings.HasPrefix(h[0], "AWS4") {
+		// Received a signed request, just pass it on.
+		return st.Transport.RoundTrip(req)
+	}
+
+	if strings.Contains(req.URL.RawPath, "%2C") {
+		// Escaping path
+		req.URL.RawPath = url.PathEscape(req.URL.RawPath)
+	}
+
+	now := time.Now().UTC()
+	req.Header.Set("Date", now.Format(time.RFC3339))
+
+	var err error
+	switch req.Body {
+	case nil:
+		_, err = st.Signer.Sign(req, nil, "es", st.Region, now)
+	default:
+		switch body := req.Body.(type) {
+		case io.ReadSeeker:
+			_, err = st.Signer.Sign(req, body, "es", st.Region, now)
+		default:
+			var buf []byte
+			buf, err = io.ReadAll(req.Body)
+			if err != nil {
+				return nil, err
+			}
+			req.Body = io.NopCloser(bytes.NewReader(buf))
+			_, err = st.Signer.Sign(req, bytes.NewReader(buf), "es", st.Region, time.Now().UTC())
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	return st.Transport.RoundTrip(req)
+}
+
+// Read method to read the content from io.reader to string
+func Read(r io.Reader) string {
+	var b bytes.Buffer
+	_, _ = b.ReadFrom(r)
+	return b.String()
+}
```

Comment:
I believe this one's only used inside the unit test code and I think swallowing the error returned by `b.ReadFrom(r)` isn't ideal. Would it be possible to remove it from here?

Reply:
I moved this method back to tests and handled err as well.